File: src/main/java/carnage/actionLogger/listeners/MiscListeners.java
package carnage.actionLogger.listeners;

import carnage.actionLogger.models.LogEntry;
import carnage.actionLogger.queue.LogQueue;
import org.bukkit.event.EventHandler;
import org.bukkit.event.Listener;
import org.bukkit.event.player.*;

public class MiscListeners implements Listener {
    private final LogQueue queue;

    public MiscListeners(LogQueue queue) { this.queue = queue; }

    @EventHandler public void onJoin(PlayerJoinEvent e) { queue.addLog(new LogEntry(e.getPlayer().getName(), "Login", "", e.getPlayer().getLocation())); }
    @EventHandler public void onQuit(PlayerQuitEvent e) { queue.addLog(new LogEntry(e.getPlayer().getName(), "Logout", "", e.getPlayer().getLocation())); }

    @EventHandler
    public void onDrop(PlayerDropItemEvent e) {
        var stack = e.getItemDrop().getItemStack();
        String detail = stack.getType() + " x" + stack.getAmount();
        queue.addLog(new LogEntry(e.getPlayer().getName(), "Item Drop", detail, e.getPlayer().getLocation()));
    }

    @EventHandler
    public void onPickup(PlayerPickupItemEvent e) {
        var stack = e.getItem().getItemStack();
        String detail = stack.getType() + " x" + stack.getAmount();
        queue.addLog(new LogEntry(e.getPlayer().getName(), "Item Pickup", detail, e.getPlayer().getLocation()));
    }
}

File: src/main/kotlin/cat/daisy/core/events/MiscListeners.kt
package cat.daisy.core.events

import cat.daisy.core.service.ActionLogService
import cat.daisy.core.service.ActionLogService.Action
import cat.daisy.core.service.ConfigService
import org.bukkit.event.EventHandler
import org.bukkit.event.EventPriority
import org.bukkit.event.Listener
import org.bukkit.event.player.PlayerDropItemEvent
import org.bukkit.event.player.PlayerJoinEvent
import org.bukkit.event.player.PlayerPickupItemEvent
import org.bukkit.event.player.PlayerQuitEvent

/**
 * Listens for miscellaneous player events: login, logout, item drop/pickup.
 */
class MiscListeners(
    private val logs: ActionLogService,
) : Listener {
    private inline fun isEnabled(): Boolean = ConfigService.getModuleSettingBoolean(ConfigService.EVENTS, "actionlogger", "enabled", true)

    @EventHandler(priority = EventPriority.MONITOR)
    fun onJoin(event: PlayerJoinEvent) {
        if (!isEnabled()) return
        if (!ConfigService.getModuleSettingBoolean(ConfigService.EVENTS, "actionlogger", "capture.login", true)) return

        logs.submit(
            ActionLogService.LogEntry(
                playerName = event.player.name,
                action = Action.LOGIN,
                detail = "",
                location = event.player.location,
            ),
        )
    }

    @EventHandler(priority = EventPriority.MONITOR)
    fun onQuit(event: PlayerQuitEvent) {
        if (!isEnabled()) return
        if (!ConfigService.getModuleSettingBoolean(ConfigService.EVENTS, "actionlogger", "capture.logout", true)) return

        logs.submit(
            ActionLogService.LogEntry(
                playerName = event.player.name,
                action = Action.LOGOUT,
                detail = "",
                location = event.player.location,
            ),
        )
    }

    @EventHandler(priority = EventPriority.MONITOR, ignoreCancelled = true)
    fun onDrop(event: PlayerDropItemEvent) {
        if (!isEnabled()) return
        if (!ConfigService.getModuleSettingBoolean(ConfigService.EVENTS, "actionlogger", "capture.item_drop", true)) return

        val itemStack = event.itemDrop.itemStack
        val detail = "${itemStack.type} x${itemStack.amount}"

        logs.submit(
            ActionLogService.LogEntry(
                playerName = event.player.name,
                action = Action.ITEM_DROP,
                detail = detail,
                location = event.player.location,
            ),
        )
    }

    @Suppress("DEPRECATION") // PlayerPickupItemEvent is deprecated but necessary for this feature
    @EventHandler(priority = EventPriority.MONITOR, ignoreCancelled = true)
    fun onPickup(event: PlayerPickupItemEvent) {
        if (!isEnabled()) return
        if (!ConfigService.getModuleSettingBoolean(ConfigService.EVENTS, "actionlogger", "capture.item_pickup", true)) return

        val itemStack = event.item.itemStack
        val detail = "${itemStack.type} x${itemStack.amount}"

        logs.submit(
            ActionLogService.LogEntry(
                playerName = event.player.name,
                action = Action.ITEM_PICKUP,
                detail = detail,
                location = event.player.location,
            ),
        )
    }
}
